Multiple Fixes for the ModPrefabCache

MrPurple6411 opened the pull request:
### Changes made in this pull request

  - Added a ModPrefabRequest Cache to call upon so that it is not making multiple requests anymore.
  - Null Check ProcessPrefabAsync
  - Add prefab name to error log when missing PrefabIdentifier

### Breaking changes

  - Should be none.

MrPurple6411 committed:
Multiple Fixes for the ModPrefabCache

1. Set cached prefabs as inactive.
2. Set check before caching prefab as its all async and may have been registered moments before.
3. Null Check ProcessPrefabAsync
4. Add prefab name to error log when missing PrefabIdentifier

MrPurple6411 committed:
Remove Double Instantiation

MrPurple6411 committed:
Set prefabs active as parent is inactive so it does not need to be.

32Kallies commented:
This seems fine for now, should be ready to merge, but it would probably be best to make a system that uses existing prefab requests for “work in progress” prefabs rather than running multiple requests at the same time.

Metious commented on Nautilus/Assets/ModPrefabCache.cs:
```diff
@@ -89,27 +90,24 @@ private void Awake()
+        else if(!Entries.ContainsKey(prefabIdentifier.classId))
         {
+            prefab.transform.parent = _prefabRoot;
+            prefab.SetActive(true);
```
While this appears to be much better than before and won't always just add an entry to the cache without much checks, I think clashing and duplicates is still possible, but it's now just "less" likely.

I suggest testing the example mod. that one has a duplicate on the copper clone iirc.

Metious commented on Nautilus/Handlers/PrefabHandler.cs:
```diff
@@ -69,7 +76,18 @@ private static IEnumerator ProcessPrefabAsync(GameObject obj, PrefabInfo info, P
+        if (ModPrefabCache.TryGetPrefabFromCache(info.ClassID, out var obj2))
+        {
+            gameObject.Set(obj2);
+            GameObject.Destroy(obj);
```
This looks dangerous. Since non-instantiated asset bundle prefabs can be custom prefabs, wouldn't destroying a prefab would make the current one also null?

MrPurple6411 replied:
> This seems fine for now, should be ready to merge, but it would probably be best to make a system that uses existing prefab requests for “work in progress” prefabs rather than running multiple requests at the same time.

> This looks dangerous. Since non-instantiated asset bundle prefabs can be custom prefabs, wouldn't destroying a prefab would make the current one also null?

Ohh ooohhhhhhhh   brain explosion..... 

Forgot that addressables Cache their request calls so as not to duplicate them.

I need to add a modPrefabRequest cache and it will fix both of these issues.

Thanx guys.

Metious reviewed (commented):
Added a few comments

MrPurple6411 committed:
Fix duplicate Cache at request level instead of just trying to destroy the duplicates.

Created a Dictionary<string, ModPrefabRequest> called Requests in tht ModPrefabCache.
Made GetModPrefabAsync return the cached request instead of making a new one.
Made ModPrefabRequest add itself to the ModPrefabCache.Requests when created.

Retargeted InstantiateAsync and SpawnAsyncReplacement to use GetModPrefabAsync isntead of bypassing the request system.

Metious reviewed (approved):
lgtm

MrPurple6411 committed:
Merge remote-tracking branch 'upstream/master' into ModPrefabCacheFixes
